File: sdks/python/src/oris_sdk/experience_v1.py
"""Generated-shape models for ``oris.experience.bundle/v1``.

The JSON Schema in ``spec/experience`` is authoritative. These dataclasses
preserve every contract field without adding an SDK dependency.
"""
from __future__ import annotations

from dataclasses import asdict, dataclass, field
from typing import Any


@dataclass
class Applicability:
    required_signals: list[str] = field(default_factory=list)
    excluded_signals: list[str] = field(default_factory=list)
    environments: list[dict[str, Any]] = field(default_factory=list)
    project_ids: list[str] = field(default_factory=list)
    tenant_ids: list[str] = field(default_factory=list)
    do_not_use_when: list[str] = field(default_factory=list)


@dataclass
class GeneV1:
    id: str
    version: int
    name: str
    description: str
    scope: str
    task_category: str
    applicability: Applicability
    steps: list[dict[str, Any]]
    tool_requirements: list[dict[str, Any]]
    safety: dict[str, Any]
    validation: dict[str, Any]
    provenance: dict[str, Any]
    lifecycle: str
    created_at: str
    updated_at: str
    metadata: dict[str, Any] = field(default_factory=dict)


@dataclass
class CapsuleV1:
    id: str
    gene_id: str
    gene_version: int
    environment_fingerprint: str
    task_context_hash: str
    execution_evidence_hash: str
    validation: dict[str, Any]
    artifact_refs: list[str]
    redaction: str
    created_at: str


@dataclass
class UsageReceiptV1:
    id: str
    gene_id: str
    gene_version: int
    agent_id: str
    run_id: str
    task_context_hash: str
    adoption: str
    applied_step_ids: list[str]
    outcome: str
    test_evidence_refs: list[str]
    created_at: str
    failure_reason: str | None = None
    cost: dict[str, Any] | None = None


@dataclass
class ExperienceBundleV1:
    schema_version: str
    gene: GeneV1
    capsules: list[CapsuleV1] = field(default_factory=list)
    usage_receipts: list[UsageReceiptV1] = field(default_factory=list)

    @classmethod
    def from_dict(cls, value: dict[str, Any]) -> "ExperienceBundleV1":
        gene = value["gene"]
        return cls(
            schema_version=value["schema_version"],
            gene=GeneV1(**{**gene, "applicability": Applicability(**gene["applicability"])}),
            capsules=[CapsuleV1(**item) for item in value.get("capsules", [])],
            usage_receipts=[UsageReceiptV1(**item) for item in value.get("usage_receipts", [])],
        )

    def to_dict(self) -> dict[str, Any]:
        return asdict(self)

    def validate(self) -> None:
        if self.schema_version != "oris.experience.bundle/v1":
            raise ValueError(f"unsupported schema_version: {self.schema_version}")
        if not self.gene.steps or not self.gene.validation.get("checks"):
            raise ValueError("a Gene requires steps and validation checks")
        if self.gene.lifecycle == "stable":
            p = self.gene.provenance
            if p.get("verified_successes", 0) < 3 or p.get("distinct_task_contexts", 0) < 2:
                raise ValueError("stable Gene lacks promotion evidence")
        for receipt in self.usage_receipts:
            if receipt.outcome == "succeeded" and receipt.adoption == "adopted" and not receipt.test_evidence_refs:
                raise ValueError("successful adoption requires test evidence")

File: sdks/typescript/src/experience-v1.ts
/** Generated-shape models for spec/experience/experience-bundle-v1.schema.json. */
export type ExperienceScope = "local" | "project" | "tenant" | "team" | "network";
export type LifecycleState = "candidate" | "stable" | "deprecated" | "quarantined" | "revoked";
export type OutcomeStatus = "succeeded" | "failed" | "safety_failed" | "inconclusive";

export interface ExperienceBundleV1 { schema_version:"oris.experience.bundle/v1"; gene:GeneV1; capsules:CapsuleV1[]; usage_receipts:UsageReceiptV1[] }
export interface GeneV1 { id:string; version:number; name:string; description:string; scope:ExperienceScope; task_category:string; applicability:Applicability; steps:ProcedureStep[]; tool_requirements:ToolRequirement[]; safety:SafetyConstraints; validation:ValidationContract; provenance:Provenance; lifecycle:LifecycleState; created_at:string; updated_at:string; metadata:Record<string,unknown> }
export interface Applicability { required_signals:string[]; excluded_signals:string[]; environments:EnvironmentConstraint[]; project_ids:string[]; tenant_ids:string[]; do_not_use_when:string[] }
export interface EnvironmentConstraint { key:string; operator:"equals"|"not_equals"|"contains"|"semver"|"exists"; value:unknown }
export interface ProcedureStep { id:string; instruction:string; tool:string|null; requires_approval:boolean; expected_output:string|null }
export interface ToolRequirement { name:string; minimum_version:string|null; required_permissions:string[] }
export interface SafetyConstraints { suggestion_only:boolean; forbidden_operations:string[]; required_approvals:string[]; secret_handling:"redact"|"reject"|"allow_references_only" }
export interface ValidationContract { checks:ValidationCheck[]; success_condition:"all"|"any"|"human_approved" }
export interface ValidationCheck { id:string; command_or_assertion:string; evidence_kind:"test"|"command"|"diff"|"artifact"|"human_approval"; timeout_seconds:number|null }
export interface Provenance { source_agent:string; source_run_id:string; trace_refs:string[]; extractor_version:string|null; verified_successes:number; verified_failures:number; distinct_task_contexts:number }
export interface CapsuleV1 { id:string; gene_id:string; gene_version:number; environment_fingerprint:string; task_context_hash:string; execution_evidence_hash:string; validation:Record<string,unknown>; artifact_refs:string[]; redaction:"redacted"|"verified_clean"|"contains_references_only"|"rejected"; created_at:string }
export interface UsageReceiptV1 { id:string; gene_id:string; gene_version:number; agent_id:string; run_id:string; task_context_hash:string; adoption:"adopted"|"partially_adopted"|"rejected"|"not_applicable"; applied_step_ids:string[]; outcome:OutcomeStatus; failure_reason?:string|null; test_evidence_refs:string[]; cost?:Record<string,unknown>|null; created_at:string }

export function validateExperienceBundleV1(bundle:ExperienceBundleV1):void {
  if(bundle.schema_version!=="oris.experience.bundle/v1") throw new Error("unsupported experience schema");
  if(!bundle.gene.steps.length||!bundle.gene.validation.checks.length) throw new Error("Gene requires steps and validation checks");
  for(const receipt of bundle.usage_receipts) if(receipt.adoption==="adopted"&&receipt.outcome==="succeeded"&&!receipt.test_evidence_refs.length) throw new Error("successful adoption requires evidence");
}
